librepo/metadata_downloader.c

```diff
@@ -421,6 +421,10 @@ lr_download_metadata(GSList *targets,
     }
 
     process_repomd_xml(targets, fd_list, paths);
+
+    g_slist_free(fd_list);
+    g_slist_free(paths);
```

Comment:
I'm afraid it isn't that simple :confused: I think it should certainly go into a separate commit and...

There's `g_slist_free()` and `g_slist_free_full()`. The former function frees only the list itself and not its contents. But the contents of the `paths` list certainly aren't freed in every case inside `process_repomd_xml()`. The descriptiors in `fd_list` also aren't always closed. I haven't really carefully considered all the possible cases, it obviously depends on the path the code takes inside the `for` loop. Maybe the errors don't normally happen, but still. To me it seems the first three `continue`s leak both fds and paths.

This code structure is really unfortunate... My idea of fixing this is to put a `cleanup:` label to the end of the `for` loop in `process_repomd_xml()`, which would contain closing the fd and freeing the path. And then instead of the `continue`s `goto cleanup`. Or something along those lines...

Reply:
The structure indeed is unfortunate but I'm not gonna fix this (at least in this PR). I tried to deal with memory / fd leaks in a better way, so please re-review it.